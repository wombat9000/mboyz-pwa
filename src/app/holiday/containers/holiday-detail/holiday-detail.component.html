<div *ngIf="holiday$ | async as holiday">
  <h1>{{holiday.name}}</h1>
  <app-post-box [holiday]="holiday"></app-post-box>
</div>




