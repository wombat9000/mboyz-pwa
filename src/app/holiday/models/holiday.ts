import {now} from 'moment';
import {DbRecord} from './DbRecord';

export interface HolidayDTO extends DbRecord {
  readonly name: string;
  readonly created: string;
}

export function newTestHoliday(id: string): HolidayDTO {
  return {
    id: id,
    name: 'someName',
    created: now().toString()
  };
}
