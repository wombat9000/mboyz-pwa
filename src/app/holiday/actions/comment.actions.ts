import {Action} from '@ngrx/store';
import {MbComment} from '../models/comment';

export const CREATE         = '[comments CommentBox] create';
export const AF_ADDED       = '[comments Firestore] added';
export const AF_MODIFIED    = '[comments Firestore] modified';
export const AF_REMOVED     = '[comments Firestore] removed';


export class Create implements Action {
  readonly type = CREATE;

  constructor(public payload: { comment: MbComment }) {
  }
}

export class AfAdded implements Action {
  readonly type = AF_ADDED;

  constructor(readonly payload: { comment: MbComment }) {
  }
}

export class AfModified implements Action {
  readonly type = AF_MODIFIED;

  constructor(readonly payload: { comment: Partial<MbComment> }) {
  }
}

export class AfRemoved implements Action {
  readonly type = AF_REMOVED;

  constructor(readonly payload: { comment: MbComment }) {
  }
}

export type CommentActions =
  Create |
  AfAdded |
  AfModified |
  AfRemoved;

