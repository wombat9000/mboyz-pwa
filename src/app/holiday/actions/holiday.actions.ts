import {Action} from '@ngrx/store';
import {Holiday} from '../holiday.service';


export const QUERY    = '[Holiday] query pizzas';
export const ADDED    = '[Holiday] added';
export const MODIFIED = '[Holiday] modified';
export const REMOVED  = '[Holiday] removed';
export const CREATE  = '[Holiday] created';

// Initial query
export class Query implements Action {
  readonly type = QUERY;
  constructor() {}
}
// AngularFire2 StateChanges
export class Added implements Action {
  readonly type = ADDED;
  constructor(public holiday: Holiday) {}
}

export class Create implements Action {
  readonly type = CREATE;
  constructor(public holiday: Holiday) {}
}

export class Modified implements Action {
  readonly type = MODIFIED;
  constructor(public holiday: Holiday) {}
}

export class Removed implements Action {
  readonly type = REMOVED;
  constructor(public holiday: Holiday) {}
}


export type HolidayActions =
  Create |
  Query |
  Added |
  Modified |
  Removed;
