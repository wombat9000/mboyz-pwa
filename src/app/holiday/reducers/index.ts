import {createFeatureSelector, createSelector} from '@ngrx/store';
import * as fromHolidays from './holiday.reducer';
import * as fromRoot from '../../reducers';


export interface HolidaysState {
  holidays: fromHolidays.State;
}

export interface State extends fromRoot.State {
  holidays: HolidaysState;
}

export const reducers = {
  holidays: fromHolidays.reducer
};

export const getHolidaysState = createFeatureSelector<HolidaysState>('holidays');

export const getHolidayEntitiesState = createSelector(
  getHolidaysState,
  state => state.holidays
);

export const getSelectedHolidayId = createSelector(
  getHolidayEntitiesState,
  fromHolidays.getSelectedId
);

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = fromHolidays.adapter.getSelectors(getHolidayEntitiesState);

export const getSelectedHoliday = createSelector(
  selectEntities,
  getSelectedHolidayId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);
