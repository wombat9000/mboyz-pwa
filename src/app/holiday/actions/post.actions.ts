import {Action} from '@ngrx/store';
import {PostDTO} from '../models/post';
import {CreateAction} from '../../core/actions/data.actions';

export const CREATE = '[posts PostboxComponent] create';
export const CREATE_SUCCESS = '[posts App] create success';
export const AF_ADDED = '[posts Firestore] added';
export const AF_MODIFIED = '[posts Firestore] modified';
export const AF_REMOVED = '[posts Firestore] removed';

export class Create implements CreateAction {
  readonly type = CREATE;

  constructor(public readonly payload: { record: PostDTO }) {
  }
}

export class CreateSuccess implements Action {
  readonly type = CREATE_SUCCESS;

  constructor() {
  }
}

export class AfAdded implements CreateAction {
  readonly type = AF_ADDED;

  constructor(public readonly payload: { record: PostDTO }) {
  }
}

export class AfModified implements Action {
  readonly type = AF_MODIFIED;

  constructor(public readonly payload: { record: Partial<PostDTO> }) {
  }
}

export class AfRemoved implements Action {
  readonly type = AF_REMOVED;

  constructor(public readonly payload: { record: PostDTO }) {
  }
}

export type PostActions =
  Create |
  AfAdded |
  AfModified |
  AfRemoved;

