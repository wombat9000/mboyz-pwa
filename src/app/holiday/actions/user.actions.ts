import {Action} from '@ngrx/store';
import {CreateAction} from '../../core/actions/data.actions';
import {MtravelUser} from '../../auth/services/auth.service';

export const CREATE = '[users PostboxComponent] create';
export const CREATE_SUCCESS = '[users App] create success';
export const AF_ADDED = '[users Firestore] added';
export const AF_MODIFIED = '[users Firestore] modified';
export const AF_REMOVED = '[users Firestore] removed';

export class Create implements CreateAction {
  readonly type = CREATE;

  constructor(public readonly payload: { record: MtravelUser }) {
  }
}

export class CreateSuccess implements Action {
  readonly type = CREATE_SUCCESS;

  constructor() {
  }
}

export class AfAdded implements CreateAction {
  readonly type = AF_ADDED;

  constructor(public readonly payload: { record: MtravelUser }) {
  }
}

export class AfModified implements Action {
  readonly type = AF_MODIFIED;

  constructor(public readonly payload: { record: Partial<MtravelUser> }) {
  }
}

export class AfRemoved implements Action {
  readonly type = AF_REMOVED;

  constructor(public readonly payload: { record: MtravelUser }) {
  }
}

export type MtravelUserActions =
  Create |
  AfAdded |
  AfModified |
  AfRemoved;

