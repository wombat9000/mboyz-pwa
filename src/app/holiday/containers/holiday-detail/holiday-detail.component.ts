import {Component, OnInit} from '@angular/core';
import {ActivatedRoute} from '@angular/router';
import * as fromRoot from '../../../reducers';
import * as fromAuth from '../../../auth/reducers';
import * as fromHoliday from '../../reducers';
import {Store} from '@ngrx/store';
import {Select} from '../../actions/holiday.actions';
import {HolidayDTO} from '../../models/holiday';
import {MtravelUser} from '../../../auth/services/auth.service';
import {Observable} from 'rxjs';
import {PostDTO} from '../../models/post';
import * as moment from 'moment';
import {map} from 'rxjs/operators';
import {SetTitle} from '../../../core/actions/app-bar.actions';

@Component({
  selector: 'app-holiday-detail',
  template: `
    <div *ngIf="holiday$ | async as holiday">
      <h1>{{holiday.name}}</h1>
      <app-post-box [holiday]="holiday"
                    [activeUser]="activeUser$ | async"
                    [posts]="posts$ | async">
      </app-post-box>
    </div>
  `,
  styles: [`
    :host {
      display: block;
      padding: 10px;
      margin: auto;
      background-color: aliceblue;
    }
  `]
})
export class HolidayDetailPageComponent implements OnInit {
  holiday$: Observable<HolidayDTO | null> = this.store.select(fromHoliday.getSelectedHoliday);
  activeUser$: Observable<MtravelUser | null> = this.store.select(fromAuth.getUser);
  posts$: Observable<PostDTO[]> = this.store.select(fromHoliday.getSelectedPosts).pipe(
    map(it => {
      return it.sort((some, other) => {
        return moment(some.created).isAfter(moment(other.created)) ? 0 : 1;
      });
    }));

  constructor(private store: Store<fromRoot.State>,
              private route: ActivatedRoute) {
  }

  ngOnInit() {
    const id: string | null = this.route.snapshot.paramMap.get('id');
    if (id !== null) {
      this.store.dispatch(new Select({id: id}));
    }

    // TODO: test
    this.store.dispatch(new SetTitle({newTitle: 'overview'}));
  }
}
