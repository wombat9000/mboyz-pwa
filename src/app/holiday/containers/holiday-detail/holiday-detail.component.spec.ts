import {async, ComponentFixture, TestBed} from '@angular/core/testing';

import {HolidayDetailPageComponent} from './holiday-detail.component';
import {CUSTOM_ELEMENTS_SCHEMA} from '@angular/core';
import {Observable} from 'rxjs/Observable';
import {By} from '@angular/platform-browser';
import {ActivatedRoute} from '@angular/router';
import {Holiday} from '../../models/holiday';
import {combineReducers, Store, StoreModule} from '@ngrx/store';
import * as fromHoliday from '../../../holiday/reducers';
import {HolidaysState} from '../../reducers';
import {Select} from '../../actions/holiday.actions';

describe('HolidayDetailComponent', () => {
  let component: HolidayDetailPageComponent;
  let fixture: ComponentFixture<HolidayDetailPageComponent>;
  let store: Store<HolidaysState>;

  const someHoliday: Holiday = {id: 'someId', name: 'someName'};

  const snapshot = {
    paramMap: new Map([
      ['id', someHoliday.id]
    ])
  };

  const activatedRoute = {
    snapshot: snapshot
  };

  beforeEach(async(() => {

    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          holidayPlanner: combineReducers(fromHoliday.reducers),
        })
      ],
      providers: [
        {provide: ActivatedRoute, useValue: activatedRoute},
      ],
      declarations: [HolidayDetailPageComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
      .compileComponents();
  }));

  beforeEach(async () => {
    store = TestBed.get(Store);
    spyOn(store, 'select').and.returnValue(Observable.of(someHoliday));
    spyOn(store, 'dispatch');

    fixture = TestBed.createComponent(HolidayDetailPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should show the holidays name', async () => {
    await fixture.whenStable();
    fixture.detectChanges();
    expect(store.dispatch).toHaveBeenCalledWith(new Select({id: someHoliday.id}));

    const heading = fixture.debugElement.query(By.css('h1')).nativeElement.textContent;
    expect(heading).toBe(someHoliday.name);
  });
});
