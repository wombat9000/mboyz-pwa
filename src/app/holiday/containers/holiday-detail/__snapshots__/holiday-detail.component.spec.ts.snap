// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HolidayDetailComponent should match snapshot: fixture 1`] = `
<app-holiday-detail
  activeUser={[Function Object]}
  holiday={[Function Object]}
  posts={[Function Array]}
  store={[Function Store]}
>
  <app-forum
    _ngcontent-c0=""
  />
</app-holiday-detail>
`;
